tools/topology2/include/components/component.conf

```diff
@@ -0,0 +1,58 @@
+# Common widget definitions
+
+#
+# Attributes
+#
+
+#
+# no_pm - maps to the DAPM widget's reg field
+# "false" value indicates that there is no direct DAPM for this widget
+#
+DefineAttribute."no_pm" {}
+
+#
+# Widget Type - maps to the widget ID with values of type enum SND_SOC_TPLG_DAPM_*
+#
+DefineAttribute.widget_type {}
+
+#
+# Stream name - maps to the widget's sname
+#
+DefineAttribute."stream_name" {}
+
+#
+# Tuple definitions added to widget's private data
+#
+
+DefineAttribute.period_sink_count {
+	# Token set reference name and type
+	token_ref	"sof_tkn_comp.word"
+}
+
+DefineAttribute.period_source_count {
+	# Token set reference name and type
+	token_ref	"sof_tkn_comp.word"
+}
+
+DefineAttribute.format {
+	# Token set reference name and type
+	token_ref	"sof_tkn_comp.string"
+	constraints {
+		values [
+			"s16_le"
+			"s24_le"
+			"s32_le"
+			"float"
+		]
+	}
+}
+
+DefineAttribute.core_id {
+	# Token set reference name and type
+	token_ref	"sof_tkn_comp.word"
```

Comment:
ok, I think I know what you mean by this. Speaking in "old terms," you want to have the compiler join "sof_tkn_comp" and "core_id" and generate a `SOF_TKN_COMP_CORE_ID` token, which will be set from this attribute. Isn't this too restrictive? Elgl DMICs currently have this:
```
SectionHWConfig."DMIC0" {

	id		"6"

	}
SectionVendorTuples."DAICONFIG.DMIC0_dmic_tuples" {
	tokens "sof_dmic_tokens"
	tuples."word" {
		SOF_TKN_INTEL_DMIC_DRIVER_VERSION	"1"
		SOF_TKN_INTEL_DMIC_CLK_MIN		"500000"
		SOF_TKN_INTEL_DMIC_CLK_MAX		"4800000"
		SOF_TKN_INTEL_DMIC_DUTY_MIN		"40"
		SOF_TKN_INTEL_DMIC_DUTY_MAX		"60"
		SOF_TKN_INTEL_DMIC_SAMPLE_RATE		"48000"
		SOF_TKN_INTEL_DMIC_FIFO_WORD_LENGTH	"32"
		SOF_TKN_INTEL_DMIC_UNMUTE_RAMP_TIME_MS	"400"

						SOF_TKN_INTEL_DMIC_NUM_PDM_ACTIVE	"1"
	}
}
SectionVendorTuples."DAICONFIG.DMIC0_pdm_tuples" {
	tokens "sof_dmic_pdm_tokens"
	tuples."short.pdm0" {
		SOF_TKN_INTEL_DMIC_PDM_CTRL_ID		"0"
		SOF_TKN_INTEL_DMIC_PDM_MIC_A_Enable	"1"
		SOF_TKN_INTEL_DMIC_PDM_MIC_B_Enable	"1"
		SOF_TKN_INTEL_DMIC_PDM_POLARITY_A	"0"
		SOF_TKN_INTEL_DMIC_PDM_POLARITY_B	"0"
		SOF_TKN_INTEL_DMIC_PDM_CLK_EDGE	"0"
		SOF_TKN_INTEL_DMIC_PDM_SKEW		"0"
	}


}
```
do we really want to define an attribute for each of them?

Reply:
> do we really want to define an attribute for each of them?

Unfortunately, I cant think of a better way than having a separate attribute for each of them. Im all ears if you have a suggestion

Reply:
yes, attributes must be defined for all parameters so that the compiler will validate each (type, value) as part of the topology build.